text/0000-impl-trait-type-aliases.md

```diff
@@ -0,0 +1,372 @@
+```
+
+`impl Trait` type aliases are useful for declaring types that are constrained by traits, but whose concrete type should be a hidden implementation detail. We can use it in place of return-position `impl Trait` as in the previous examples.
+
+```rust
+trait Trait {}
+
+type Baz = impl Trait;
+
+// The same as `fn bar() -> impl Baz`
+fn bar() -> Baz {
+    // ...
+}
+```
+
+However, if we use `Baz` in multiple locations, we constrain the concrete type referred to by `Baz` to be the same, so we get a type that we know will be the same everywhere and will satisfy specific bounds, whose concrete type is hidden. This can be very useful in libraries where you want to hide implementation details.
+
+```rust
+trait Trait {}
+
+type Baz = impl Trait;
+
+fn foo(x: Baz, y: Baz) {
+    // ...
+}
+
+struct Foo {
+    a: Baz,
+    b: (Baz, Baz),
+}
+```
+
+In this example, the concrete type referred to by `Baz` is guaranteed to be the same wherever `Baz` occurs.
+
+Note that using `Baz` as an argument type is *not* the same as argument-position `impl Trait`, as `Baz` refers to a unique type, whereas the concrete type for argument-position `impl Trait` is determined by the caller.
+
+```rust
+trait Trait {}
+
+type Baz = impl Trait;
+
+fn foo(x: Baz) {
+    // ...
+}
+
+// is *not* the same as:
+
+fn foo(x: impl Trait) {
+    // ...
+}
+```
+
+Just like with any other type alias, we can use `impl Trait` to specify associated types for traits, as in the following example.
+
+```rust
+trait Trait {
+    type Assoc;
+}
+
+struct Foo {}
+
+impl Trait for Foo {
+    type Assoc = impl Debug;
+}
+```
+
+Here, anything that makes use of `Foo` knows that `Foo::Assoc` implements `Debug`, but has no knowledge of its concrete type.
+
+[Eventually, we would also describe the use of `impl Trait` in `let`, `const` and `static` bindings, but as they are as-yet unimplemented and function the same as return-type `impl Trait`, they haven't been included here.]
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Since RFC 2071 was accepted, the initial implementation of `existential type` [has already been completed](https://github.com/rust-lang/rust/pull/52024). This RFC would simply replace the syntax of `existential type`, from:
+
+```rust
+existential type Foo: Bar;
+```
+
+to:
+
+```rust
+type Foo = impl Bar;
+```
+
+In addition, when documenting `impl Trait`, explanations of the feature would avoid type theoretic terminology (specifically "existential types") and prefer type inference language (if any technical description is needed at all).
+
+## Restricting compound `impl Trait` trait aliases
+The type alias syntax is more flexible than `existential type`, but for now we restrict the form to that equivalent to `existential type`. That means that, if `impl Trait` appears on the right-hand side of a type alias declaration, it must be the only type. The following compound type aliases, therefore, are initially forbidden:
```

Comment:
My biggest issue with this restriction is that it makes `impl Trait` inconsistent between type aliases and everywhere  else (excluding the already inconsistent argument position). With `existential type` there was a simple rule that could be applied to `impl Trait` in every position except argument position: it introduces a new anonymous `existential type` in the current context. This rule works perfectly for return position, type declaration in bindings, type aliases, and could work for type declaration of struct/enum members if there wasn't a chance of confusion with argument-position-impl-trait.

Reply:
I'm not quite sure I follow your point. This restriction is simply a syntactic one: it is simply intended to sidestep the question of what:
```rust
type Foo = (impl Bar, impl Bar);
``` 
means for now (because some people expressed unease at this construction in particular).

`impl Trait` continues to be applicable in exactly the same places as `existential type`: this rule simply means it can't be used in more complex scenarios than `existential type` yet.

Reply:
I mean that with existential types it's possible to have a single very simple rule for desugaring `impl Trait` that covers both:

```rust
type Foo = (impl Bar, impl Bar);
let foo: (impl Bar, impl Bar);
```

but with `type Foo = impl Trait;` trying to apply the same sort of rule you get to this recursive definition that needs a special case when you use a bare `impl Trait` in a type alias.

Reply:
Ah, I see, yes. You do have to have a single type alias as a base case if you're using the `impl Trait` type aliases to desugar. In practice, the desugaring of `existential type` is effectively replaced by the type alias. That is, the `existential type` design was originally intended to act as a desugaring for return-position and variable-binding (e.g. `let`) `impl Trait`. Using type aliases fills that role: but you can't use it to desugar itself. In practice, I don't think this is important, as there's no practical difference between the `existential type` itself and its alias.

Reply:
Why can't it be used to desugar, @varkor?

Reply:
Because if you try to desugar each occurrence of `impl Trait` you would end up trying to desugar:
```rust
type Foo = impl Trait;
```
into:
```rust
type Foo = impl Trait;
```
so you need to have this as a base case.